src/pudl/analysis/mcoe.py

```diff
@@ -38,7 +39,136 @@
 """
 
 
-def heat_rate_by_unit(pudl_out):
+def mcoe_asset_factory(
+    freq: Literal["AS", "MS"],
+    io_manager_key: str | None = None,
+) -> list[AssetsDefinition]:
+    """Build MCOE related assets at yearly and monthly frequencies."""
+    agg_freqs = {"AS": "yearly", "MS": "monthly"}
+    if freq not in agg_freqs:
+        raise ValueError(f"freq must be one of {agg_freqs.keys()}, got: {freq}.")
+
+    @asset(
+        name=f"heat_rate_by_unit_{agg_freqs[freq]}",
+        ins={
+            "gen": AssetIn(
+                key=f"generation_fuel_by_generator_{agg_freqs[freq]}_eia923"
+            ),
+            "bf": AssetIn(key=f"denorm_boiler_fuel_{agg_freqs[freq]}_eia923"),
+        },
+        io_manager_key=io_manager_key,
+        compute_kind="Python",
+    )
+    def hr_by_unit_asset(gen: pd.DataFrame, bf: pd.DataFrame) -> pd.DataFrame:
+        return heat_rate_by_unit(gen=gen, bf=bf)
```

Comment:
Previously we were **not** using the allocated generation fuel inputs to the MCOE calculation -- just the directly reported per-generator generation from the `generation_eia923` table, so this new output will potentially have **very different outputs** (and hopefully also much higher coverage). What kind of sanity checking have we done to make sure that the new heat rates look correct?

We should at least make some histograms of the heat rates so we can see whether the combined cycle natural gas plants, the gas combustion turbines, and the coal-fired steam turbines have the expected distribution of heat rates.

We also previously filtered out the obviously ridiculously wrong values -- e.g. cases where net generation was negative due to low capacity factors + parasitic loads, resulting in negative fuel costs, etc.

What outputs do we want to be writing into the database now?  Should we be filtering out the junk results and if so how?

Reply:
one thing to note is that i think the inputs to the mcoe as *slightly* updated from the net gen/fuel allocation stuff. the net generation side is definitely incorporated into this equation but the allocated fuel *is not* incorporated. The allocated generation was easy to insert into mcoe as an option because we had a pudl table that was the same "shape" (i.e. it had date/plant id/gen id as primary keys that we could swap into the same methodology). There was no date/plant id/gen id level table on the fuel side. so we kept this `heat_rate_by_unit` step which takes the `bf` table which is not the allocated fuel. so the fuel side of the whole mcoe dag is still the reported values.

Reply:
OHHH, you're right.  So we're not really updating the calculation much if at all here, but it would be easy to do so if we wanted to, and then this asset would only depend on the `gen` dataframe, right? And we'd be using both the generation and the fuel consumed within each record of that dataframe. Do we have reason to believe that this would be a lower quality output?  Would it deal with the multi-fuel units worse?

I guess this means that we still aren't getting a boost in the overall MCOE output coverage.